lib/kb_Metrics/metricsdb_controller.py

```diff
@@ -26,279 +26,189 @@ def _config_str_to_list(self, list_str):
+                             'invoke this action.')
 
-        params = self.process_parameters(params)
-        # TODO: set the param['minTime'] and param['maxTime'] to a given time window,
-        # such as most current 24 hours instead of 48 hours as for others.
+        params = self._process_parameters(params)
         auth2_ret = self.metrics_dbi.aggr_user_details(
             params['user_ids'], params['minTime'], params['maxTime'])
-        updData = 0
+        upDated = 0
+        upSerted = 0
         if len(auth2_ret) == 0:
             print("No user records returned for update!")
-            return updData
+            return (0, 0)
 
-        print('Retrieved {} user record(s)'.format(len(auth2_ret)))
+        print('Retrieved {} user record(s) for update!'.format(len(auth2_ret)))
         idKeys = ['username', 'email']
         dataKeys = ['full_name', 'signup_at', 'last_signin_at', 'roles']
         for u_data in auth2_ret:
-            filterByKey = lambda keys: {x: u_data[x] for x in keys}
-            idData = filterByKey(idKeys)
-            userData = filterByKey(dataKeys)
-            isKbstaff = 1 if idData['username'] in self.kbstaffList else 0
+            idData = {x: u_data[x] for x in idKeys}
+            userData = {x: u_data[x] for x in dataKeys}
+            isKbstaff = 1 if self._is_kbstaff(idData['username']) else 0
             update_ret = self.metrics_dbi.update_user_records(
                 idData, userData, isKbstaff)
-        updData += update_ret.raw_result['nModified']
-
-        return updData
-
-    def update_daily_activities(self, requesting_user, params, token):
+            if update_ret.raw_result['updatedExisting']:
+                upDated += update_ret.raw_result['nModified']
+            elif update_ret.raw_result.get('upserted'):
+                upSerted += 1
+        print('updated {} and upserted {} users.'.format(upDated, upSerted))
+        return upDated + upSerted
+
+    def _update_daily_activities(self, requesting_user, params, token):
         """
         update user activities reported from Workspace.
         If match not found, insert that record as new.
         """
-        if not self.is_metrics_admin(requesting_user):
-            raise ValueError('You do not have permission to invoke this action.')
+        if not self._is_metrics_admin(requesting_user):
+            raise ValueError('You do not have permission to '
+                             'invoke this action.')
 
-        ws_ret = self.get_activities_from_wsobjs(requesting_user, params, token)
+        ws_ret = self._get_activities_from_wsobjs(requesting_user,
+                                                  params, token)
         act_list = ws_ret['metrics_result']
-        updData = 0
+        upDated = 0
+        upSerted = 0
         if len(act_list) == 0:
             print("No activity records returned for update!")
-            return updData
+            return (0, 0)
```

Comment:
Add a test for this as well. It should return 0 or (0, 0)?

Reply:
added test to return 0